### ISSUE TITLE
FileNotFoundError: [Errno 2] No such file or directory: 'cdhit.out.clstr'

### ISSUE THREAD

**Issue by philge** [marcelauliano/MitoHiFi#9]
Hi,

I get this error when I run MitoHifi with test data.
"isNeg? FalseFiltering and writing...

1 out of 3 sequences remained after filtering.
isNeg? FalseFiltering and writing...

1 out of 3 sequences remained after filtering.
Traceback (most recent call last):
  File "/bin/mitohifi_v2.py", line 334, in <module>
    main()
  File "/bin/mitohifi_v2.py", line 283, in main
    repr_contig_id, repr_contig_cluster = getReprContig.get_repr_contig("all_mitogenomes.rotated.fa", args.t)
  File "/bin/getReprContig.py", line 64, in get_repr_contig
    repr_contig_cluster, repr_contig_info = get_largest_cluster(cdhit_out_clstr)
  File "/bin/getReprContig.py", line 33, in get_largest_cluster
    with open(cdhit_clstr_file, "r") as f:
FileNotFoundError: [Errno 2] No such file or directory: 'cdhit.out.clstr'
3 Oct).
It tends to insert more gaps into gap-rich regions than previous versions.
To disable this change, add the --leavegappyregion option."

Can you please help me to fix the issue?

Thanks

**Comment by jgnunes** [comment]
Hi @philge could you please share the complete log file here? Files tree and/or the execution command could be useful as well.

**Comment by philge** [comment]
Hi Jgnunes,

This is the command I used : singularity exec --pwd /home/ubuntu/mitohifi/test/ --bind /data_disk/software/MitoHiFi/:/data/ mitohifi-v2.sif  mitohifi_v2.py -c /data/exampleFiles/test.fa -f /data/exampleFiles/NC_016067.1.fasta -g /data/exampleFiles/NC_016067.1.gb -t 90 -o 5 1>mitohifi1.log 2>mitohifi2.log

and here is the complete log,
Building a new DB, current time: 09/01/2021 09:20:20
New DB name:   /data/exampleFiles/NC_016067.1.fasta
New DB title:  /data/exampleFiles/NC_016067.1.fasta
Sequence type: Nucleotide
Deleted existing Nucleotide BLAST database named /data/exampleFiles/NC_016067.1.fasta
Keep MBits: T
Maximum file size: 1000000000B
Adding sequences from FASTA; added 1 sequences in 0.009269 seconds.

Command line: /bin/MitoFinder/mitofinder --max-contig-size 78295 -j tig00007550_1.annotation -a tig00007550_1.mitogenome.fa -r /data/exampleFiles/NC_016067.1.gb -o 5 -p 90

Now running MitoFinder ...

Start time : 2021-09-01 09:20:27

Job name = tig00007550_1.annotation

Creating Output directory : /home/ubuntu/mitohifi/test/tig00007550_1.annotation
All results will be written here

Program folders:
MEGAHIT = /bin/MitoFinder/megahit/
MetaSPAdes folder = /bin/MitoFinder/metaspades/bin/
IDBA-UD folder = /bin/MitoFinder/idba/bin/
Blast folder = /bin/MitoFinder/blast/bin/
ARWEN folder = /bin/MitoFinder/arwen/
MiTFi folder = /bin/MitoFinder/mitfi/
tRNAscan-SE folder = /bin/MitoFinder/trnascanSE/tRNAscan-SE-2.0/


Formatting database for mitochondrial contigs identification...
Running mitochondrial contigs identification step...

MitoFinder found a single mitochondrial contig
Checking resulting contig for circularization...

Evidences of circularization could not be found, but everyother step was successful



Creating summary statistics for the mtDNA contig

Annotating mitochondrial contig

tRNA annotation with MitFi run well.

Annotation completed


Creating GFF and fasta files.

Note: 
15 genes were found in mtDNA_contig

## Final sequence saved to /home/ubuntu/mitohifi/test/tig00007550_1.annotation/tig00007550_1.annotation_MitoFinder_mitfi_Final_Results/
Total wall-clock time used by MitoFinder = 0:02:25.126726

Command line: /bin/MitoFinder/mitofinder --max-contig-size 78295 -j tig00007572_1.annotation -a tig00007572_1.mitogenome.fa -r /data/exampleFiles/NC_016067.1.gb -o 5 -p 90

Now running MitoFinder ...

Start time : 2021-09-01 09:22:54

Job name = tig00007572_1.annotation

Creating Output directory : /home/ubuntu/mitohifi/test/tig00007572_1.annotation
All results will be written here

Program folders:
MEGAHIT = /bin/MitoFinder/megahit/
MetaSPAdes folder = /bin/MitoFinder/metaspades/bin/
IDBA-UD folder = /bin/MitoFinder/idba/bin/
Blast folder = /bin/MitoFinder/blast/bin/
ARWEN folder = /bin/MitoFinder/arwen/
MiTFi folder = /bin/MitoFinder/mitfi/
tRNAscan-SE folder = /bin/MitoFinder/trnascanSE/tRNAscan-SE-2.0/


Formatting database for mitochondrial contigs identification...
Running mitochondrial contigs identification step...

MitoFinder found a single mitochondrial contig
Checking resulting contig for circularization...

Evidences of circularization could not be found, but everyother step was successful



Creating summary statistics for the mtDNA contig

Annotating mitochondrial contig

tRNA annotation with MitFi run well.

Annotation completed


Creating GFF and fasta files.

Note: 
15 genes were found in mtDNA_contig

## Final sequence saved to /home/ubuntu/mitohifi/test/tig00007572_1.annotation/tig00007572_1.annotation_MitoFinder_mitfi_Final_Results/
Total wall-clock time used by MitoFinder = 0:02:24.449668
================================================================
Program: CD-HIT, V4.7 (+OpenMP), Jul 01 2017, 08:43:07
Command: cd-hit-est -i all_mitogenomes.rotated.fa -d 0 -c 0.8
         -n 4 -o cdhit.out -T 90 -M 2000

Started: Wed Sep  1 09:25:18 2021
================================================================
                            Output                              
----------------------------------------------------------------
		====== CD-HIT version 4.7 (built on Jul  1 2017) ======

Usage: cd-hit-est [Options] 

Options

   -i	input filename in fasta format, required
   -j	input filename in fasta/fastq format for R2 reads if input are paired end (PE) files
 	 -i R1.fq -j R2.fq -o output_R1 -op output_R2 or
 	 -i R1.fa -j R2.fa -o output_R1 -op output_R2 
   -o	output filename, required
   -op	output filename for R2 reads if input are paired end (PE) files
   -c	sequence identity threshold, default 0.9
 	this is the default cd-hit's "global sequence identity" calculated as:
 	number of identical amino acids in alignment
 	divided by the full length of the shorter sequence
   -G	use global sequence identity, default 1
 	if set to 0, then use local sequence identity, calculated as :
 	number of identical amino acids in alignment
 	divided by the length of the alignment
 	NOTE!!! don't use -G 0 unless you use alignment coverage controls
 	see options -aL, -AL, -aS, -AS
   -b	band_width of alignment, default 20
   -M	memory limit (in MB) for the program, default 800; 0 for unlimitted;
   -T	number of threads, default 1; with 0, all CPUs will be used
   -n	word_length, default 10, see user's guide for choosing it
   -l	length of throw_away_sequences, default 10
   -d	length of description in .clstr file, default 20
 	if set to 0, it takes the fasta defline and stops at first space
   -s	length difference cutoff, default 0.0
 	if set to 0.9, the shorter sequences need to be
 	at least 90% length of the representative of the cluster
   -S	length difference cutoff in amino acid, default 999999
 	if set to 60, the length difference between the shorter sequences
 	and the representative of the cluster can not be bigger than 60
   -aL	alignment coverage for the longer sequence, default 0.0
 	if set to 0.9, the alignment must covers 90% of the sequence
   -AL	alignment coverage control for the longer sequence, default 99999999
 	if set to 60, and the length of the sequence is 400,
 	then the alignment must be >= 340 (400-60) residues
   -aS	alignment coverage for the shorter sequence, default 0.0
 	if set to 0.9, the alignment must covers 90% of the sequence
   -AS	alignment coverage control for the shorter sequence, default 99999999
 	if set to 60, and the length of the sequence is 400,
 	then the alignment must be >= 340 (400-60) residues
   -A	minimal alignment coverage control for the both sequences, default 0
 	alignment must cover >= this value for both sequences 
   -uL	maximum unmatched percentage for the longer sequence, default 1.0
 	if set to 0.1, the unmatched region (excluding leading and tailing gaps)
 	must not be more than 10% of the sequence
   -uS	maximum unmatched percentage for the shorter sequence, default 1.0
 	if set to 0.1, the unmatched region (excluding leading and tailing gaps)
 	must not be more than 10% of the sequence
   -U	maximum unmatched length, default 99999999
 	if set to 10, the unmatched region (excluding leading and tailing gaps)
 	must not be more than 10 bases
   -B	1 or 0, default 0, by default, sequences are stored in RAM
 	if set to 1, sequence are stored on hard drive
 	!! No longer supported !!
   -P	input paired end (PE) reads, default 0, single file
 	if set to 1, please use -i R1 -j R2 to input both PE files
   -cx	length to keep after trimming the tail of sequence, default 0, not trimming
 	if set to 50, the program only uses the first 50 letters of input sequence
   -cy	length to keep after trimming the tail of R2 sequence, default 0, not trimming
 	if set to 50, the program only uses the first 50 letters of input R2 sequence
 	e.g. -cx 100 -cy 80 for paired end reads
   -ap	alignment position constrains,  default 0, no constrain
 	if set to 1, the program will force sequences to align at beginings
 	when set to 1, the program only does +/+ alignment
   -p	1 or 0, default 0
 	if set to 1, print alignment overlap in .clstr file
   -g	1 or 0, default 0
 	by cd-hit's default algorithm, a sequence is clustered to the first 
 	cluster that meet the threshold (fast cluster). If set to 1, the program
 	will cluster it into the most similar cluster that meet the threshold
 	(accurate but slow mode)
 	but either 1 or 0 won't change the representatives of final clusters
   -r	1 or 0, default 1, by default do both +/+ & +/- alignments
 	if set to 0, only +/+ strand alignment
   -mask	masking letters (e.g. -mask NX, to mask out both 'N' and 'X')
   -match	matching score, default 2 (1 for T-U and N-N)
   -mismatch	mismatching score, default -2
   -gap	gap opening score, default -6
   -gap-ext	gap extension score, default -1
   -bak	write backup cluster file (1 or 0, default 0)
   -sc	sort clusters by size (number of sequences), default 0, output clusters by decreasing length
 	if set to 1, output clusters by decreasing size
   -sf	sort fasta/fastq by cluster size (number of sequences), default 0, no sorting
 	if set to 1, output sequences by decreasing cluster size
   -h	print this help

   Questions, bugs, contact Limin Fu at l2fu@ucsd.edu, or Weizhong Li at liwz@sdsc.edu
   For updated versions and information, please visit: http://cd-hit.org

   cd-hit web server is also available from http://cd-hit.org

   If you find cd-hit useful, please kindly cite:

   "CD-HIT: a fast program for clustering and comparing large sets of protein or nucleotide sequences", Weizhong Li & Adam Godzik. Bioinformatics, (2006) 22:1658-1659
   "CD-HIT: accelerated for clustering the next generation sequencing data", Limin Fu, Beifang Niu, Zhengwei Zhu, Sitao Wu & Weizhong Li. Bioinformatics, (2012) 28:3150-3152


MitoHifi v2.2/n
Started at: 2021-09-01
Length of related mitogenome is: 15659 bp

Running MitoHifi pipeline in contigs mode

Fixing potentially conflicting FASTA headers...


Let's run the blast of the contigs versus the close-related mitogenome

makeblastdb -in /data/exampleFiles/NC_016067.1.fasta -dbtype nucl

makeblastdb done. Running blast with the contigs

Blast done!


6-) Now we are going to circularize, annotate and rotate each contig which is a potential mitogenome


Working with contig: tig00007550_1

Running mitofinder with maximum contig size of 78295 bp
Rotated to tRNA-Phe genome is at tig00007550_1.mitogenome.rotated.fa

Working with contig: tig00007572_1

Running mitofinder with maximum contig size of 78295 bp
Rotated to tRNA-Phe genome is at tig00007572_1.mitogenome.rotated.fa

7-) Now the final rotated contigs will be aligned

List of contigs that will be aligned: ['tig00007572_1.mitogenome.rotated.fa', 'tig00007550_1.mitogenome.rotated.fa']

MAFFT output will be in clustal format
MAFFT alignment will be called with:
mafft --clustalout --thread 90 all_mitogenomes.rotated.fa > all_mitogenomes.rotated.aligned.fa


8-) Now we will choose the most representative contig
isNeg? FalseFiltering and writing...

1 out of 3 sequences remained after filtering.
isNeg? FalseFiltering and writing...

1 out of 3 sequences remained after filtering.
Traceback (most recent call last):
File "/bin/mitohifi_v2.py", line 334, in
main()
File "/bin/mitohifi_v2.py", line 283, in main
repr_contig_id, repr_contig_cluster = getReprContig.get_repr_contig("all_mitogenomes.rotated.fa", args.t)
File "/bin/getReprContig.py", line 64, in get_repr_contig
repr_contig_cluster, repr_contig_info = get_largest_cluster(cdhit_out_clstr)
File "/bin/getReprContig.py", line 33, in get_largest_cluster
with open(cdhit_clstr_file, "r") as f:
FileNotFoundError: [Errno 2] No such file or directory: 'cdhit.out.clstr'
3 Oct).
It tends to insert more gaps into gap-rich regions than previous versions.
To disable this change, add the --leavegappyregion option.

Thanks

**Comment by jgnunes** [comment]
Hi @philge, thanks for sharing the log. Is this the content of both logs (mitohifi1.log and mitohifi2.log)? If not, could you share both? Also, can you confirm if the pipeline has successfully created the `all_mitogenomes.rotated.fa` and `all_mitogenomes.rotated.aligned.fa` files?  

Best,

**Comment by philge** [comment]
> Hi @philge, thanks for sharing the log. Is this the content of both logs (mitohifi1.log and mitohifi2.log)? If not, could you share both? Also, can you confirm if the pipeline has successfully created the `all_mitogenomes.rotated.fa` and `all_mitogenomes.rotated.aligned.fa` files?
> 
> Best,

Yes, it is the content of both logs. Pipeline has successfully created the `all_mitogenomes.rotated.fa` and `all_mitogenomes.rotated.aligned.fa` files.

**Comment by jgnunes** [comment]
It seems like some usage error with cd-hit. Interestingly, your cd-hit command (`cd-hit-est -i all_mitogenomes.rotated.fa -d 0 -c 0.8 -n 4 -o cdhit.out -T 90 -M 2000`) uses a `-M` option which is not being used in MitoHiFi's current version (which should call `cd-hit-est -i all_mitogenomes.rotated.fa -d 0 -c 0.8 -n 4 -o cdhit.out -T 90`). Have you modified anything from MitoHiFi's original code or maybe have you run cd-hit-est separately?  

Also, your cd-hit version (4.7) is outdated: MitoHiFi has been tested with the newer version 4.8. Maybe updating cd-hit will solve that problem. Anyways, I've failed to reproduce your issue using version 4.7 and the same command (`cd-hit-est -i all_mitogenomes.rotated.fa -d 0 -c 0.8 -n 4 -o cdhit.out -T 90 -M 2000`): cd-hit has run successfully.

To sum it up, I would try to update cdhit to 4.8. If that does not help, can you share the working directory content with me? I want to make sure all the intermediate files are being created.

**Comment by philge** [comment]
> It seems like some usage error with cd-hit. Interestingly, your cd-hit command (`cd-hit-est -i all_mitogenomes.rotated.fa -d 0 -c 0.8 -n 4 -o cdhit.out -T 90 -M 2000`) uses a `-M` option which is not being used in MitoHiFi's current version (which should call `cd-hit-est -i all_mitogenomes.rotated.fa -d 0 -c 0.8 -n 4 -o cdhit.out -T 90`). Have you modified anything from MitoHiFi's original code or maybe have you run cd-hit-est separately?
> 
> Also, your cd-hit version (4.7) is outdated: MitoHiFi has been tested with the newer version 4.8. Maybe updating cd-hit will solve that problem. Anyways, I've failed to reproduce your issue using version 4.7 and the same command (`cd-hit-est -i all_mitogenomes.rotated.fa -d 0 -c 0.8 -n 4 -o cdhit.out -T 90 -M 2000`): cd-hit has run successfully.
> 
> To sum it up, I would try to update cdhit to 4.8. If that does not help, can you share the working directory content with me? I want to make sure all the intermediate files are being created.

I have added "-M 2000" coz I got a memory error. I am using [cd-hit](url) version installed by Docker. Here is the directory content,
-rw-rw-r-- 1 ubuntu ubuntu    124 Sep  1 09:22 all_contigs.circularisationCheck.txt
-rw-rw-r-- 1 ubuntu ubuntu  60045 Sep  1 09:25 all_mitogenomes.rotated.aligned.fa
-rw-rw-r-- 1 ubuntu ubuntu  31605 Sep  1 09:25 all_mitogenomes.rotated.fa
-rw-rw-r-- 1 ubuntu ubuntu  13925 Sep  1 09:22 circularization_check.blast.tsv
-rw-rw-r-- 1 ubuntu ubuntu 186837 Sep  1 09:22 circularization_check.blast.xml
-rw-rw-r-- 1 ubuntu ubuntu   1356 Sep  1 09:20 contigs.blastn
-rw-rw-r-- 1 ubuntu ubuntu     28 Sep  1 09:20 contigs_ids.txt
-rw-rw-r-- 1 ubuntu ubuntu    172 Sep  1 09:20 parsed_blast.txt
-rw-rw-r-- 1 ubuntu ubuntu    172 Sep  1 09:20 parsed_blast_all.txt
drwxrwxr-x 5 ubuntu ubuntu   4096 Sep  1 09:22 tig00007550_1.annotation
-rw-rw-r-- 1 ubuntu ubuntu   1385 Sep  1 09:22 tig00007550_1.annotation_MitoFinder.log
-rw-rw-r-- 1 ubuntu ubuntu     32 Sep  1 09:20 tig00007550_1.circularisationCheck.txt
-rw-rw-r-- 1 ubuntu ubuntu  31598 Sep  1 09:20 tig00007550_1.mito.fa
-rw-rw-r-- 1 ubuntu ubuntu     76 Sep  1 09:20 tig00007550_1.mito.fa.nhr
-rw-rw-r-- 1 ubuntu ubuntu    104 Sep  1 09:20 tig00007550_1.mito.fa.nin
-rw-rw-r-- 1 ubuntu ubuntu   7768 Sep  1 09:20 tig00007550_1.mito.fa.nsq
-rw-rw-r-- 1 ubuntu ubuntu  15781 Sep  1 09:20 tig00007550_1.mitogenome.fa
-rw-rw-r-- 1 ubuntu ubuntu     76 Sep  1 09:20 tig00007550_1.mitogenome.fa.nhr
-rw-rw-r-- 1 ubuntu ubuntu    104 Sep  1 09:20 tig00007550_1.mitogenome.fa.nin
-rw-rw-r-- 1 ubuntu ubuntu   3878 Sep  1 09:20 tig00007550_1.mitogenome.fa.nsq
-rw-rw-r-- 1 ubuntu ubuntu  15803 Sep  1 09:22 tig00007550_1.mitogenome.rotated.fa
drwxrwxr-x 5 ubuntu ubuntu   4096 Sep  1 09:25 tig00007572_1.annotation
-rw-rw-r-- 1 ubuntu ubuntu   1385 Sep  1 09:25 tig00007572_1.annotation_MitoFinder.log
-rw-rw-r-- 1 ubuntu ubuntu     30 Sep  1 09:22 tig00007572_1.circularisationCheck.txt
-rw-rw-r-- 1 ubuntu ubuntu  16291 Sep  1 09:22 tig00007572_1.mito.fa
-rw-rw-r-- 1 ubuntu ubuntu     76 Sep  1 09:22 tig00007572_1.mito.fa.nhr
-rw-rw-r-- 1 ubuntu ubuntu    104 Sep  1 09:22 tig00007572_1.mito.fa.nin
-rw-rw-r-- 1 ubuntu ubuntu   4004 Sep  1 09:22 tig00007572_1.mito.fa.nsq
-rw-rw-r-- 1 ubuntu ubuntu  15780 Sep  1 09:22 tig00007572_1.mitogenome.fa
-rw-rw-r-- 1 ubuntu ubuntu     76 Sep  1 09:22 tig00007572_1.mitogenome.fa.nhr
-rw-rw-r-- 1 ubuntu ubuntu    104 Sep  1 09:22 tig00007572_1.mitogenome.fa.nin
-rw-rw-r-- 1 ubuntu ubuntu   3878 Sep  1 09:22 tig00007572_1.mitogenome.fa.nsq
-rw-rw-r-- 1 ubuntu ubuntu  15802 Sep  1 09:25 tig00007572_1.mitogenome.rotated.fa

**Comment by jgnunes** [comment]
Hi @philge before trying to update CD-HIT, can you check if the current version runs successfully by running it with `cd-hit-est -i all_mitogenomes.rotated.fa -d 0 -c 0.8 -n 4 -o cdhit.out -T 90 -M 2000` (don't need to re-run MitoHiFi, just use the `all_mitogenomes.rotated.fa` that you already have)? Also, can you share your modified version of the `getReprContig.py`, so that I could try to reproduce the issue on my side? 

Best,

**Comment by philge** [comment]
> Hi @philge before trying to update CD-HIT, can you check if the current version runs successfully by running it with `cd-hit-est -i all_mitogenomes.rotated.fa -d 0 -c 0.8 -n 4 -o cdhit.out -T 90 -M 2000` (don't need to re-run MitoHiFi, just use the `all_mitogenomes.rotated.fa` that you already have)? Also, can you share your modified version of the `getReprContig.py`, so that I could try to reproduce the issue on my side?
> 
> Best,

I have run following command,
"time singularity exec --pwd /home/ubuntu/mitohifi/test/ --bind /data_disk/software/MitoHiFi/:/data/ mitohifi-v2.sif cd-hit-est -i /data/all_mitogenomes.rotated.fa -d 0 -c 0.8 -n 4 -o cdhit.out -T 90 -M 2000"

I think the program run. Below is the log,
================================================================
Program: CD-HIT, V4.7 (+OpenMP), Jul 01 2017, 08:43:07
Command: /usr/bin/cd-hit-est -i
         /data/all_mitogenomes.rotated.fa -d 0 -c 0.8 -n 4 -o
         cdhit.out -T 90 -M 2000

Started: Mon Sep 13 06:27:32 2021
================================================================
                            Output
----------------------------------------------------------------
total seq: 2
longest and shortest : 15507 and 15506
Total letters: 31013
Sequences have been sorted

Approximated minimal memory consumption:
Sequence        : 0M
Buffer          : 90 X 15M = 1381M
Table           : 2 X 0M = 0M
Miscellaneous   : 0M
Total           : 1381M

Table limit with the given memory limit:
Max number of representatives: 40000
Max number of word counting entries: 77367647

# comparing sequences from          0  to          2
---------- new table with        1 representatives

        2  finished          1  clusters

Apprixmated maximum memory consumption: 1381M
writing new database
writing clustering information
program completed !

Total CPU time 2.23

Here is getReprContig.py 
[getReprContig.txt](https://github.com/marcelauliano/MitoHiFi/files/7152019/getReprContig.txt)
